File: src/sharing/link-exchange.ts
import { SyncKitError } from "../core/errors.js";
import { base64UrlToBytes, bytesToBase64Url } from "../crypto/index.js";
import {
  parseSharingInvitationV1,
  parseSharingPublicKeyResponseV1,
  type SharingInvitationV1,
  type SharingPublicKeyResponseV1,
  type SharingRole,
} from "./index.js";
import {
  SHARING_JOIN_EXCHANGE_PARAM,
  SHARING_JOIN_FOLDER_PARAM,
  SHARING_JOIN_MARKER_PARAM,
} from "./join.js";

/**
 * Link-carried key exchange. The signed invitation and key-response travel in
 * the join/response links (base64url of the existing structs) instead of Drive
 * `exchanges/` files, so a `drive.file` recipient never has to read a file the
 * owner authored and the owner never has to read a file the recipient authored.
 * The only Drive object either side reads is the encrypted dataset, which the
 * recipient grants by picking the specific file(s) — see
 * docs/sync-kit-linkbased-join.md.
 */

/** A dataset file the recipient is granted, so the Picker can offer it by id. */
export type SharingDatasetFileV1 = {
  datasetId: string;
  fileId: string;
  role: Exclude<SharingRole, "owner">;
};

export const SHARING_JOIN_INVITATION_PARAM = "sk-inv";
export const SHARING_JOIN_FILES_PARAM = "sk-files";
export const SHARING_RESPONSE_MARKER_PARAM = "sk-resp";
export const SHARING_RESPONSE_PAYLOAD_PARAM = "sk-kr";

export type SharingJoinLinkV1 = {
  invitation: SharingInvitationV1;
  files: SharingDatasetFileV1[];
};

export type SharingResponseLinkV1 = {
  response: SharingPublicKeyResponseV1;
};

// --- payload encode/decode (signed structs are unchanged; only transported) ---

export function encodeSharingInvitationV1(
  invitation: SharingInvitationV1,
): string {
  return encodeJson(parseSharingInvitationV1(invitation));
}

export function decodeSharingInvitationV1(encoded: string): SharingInvitationV1 {
  return parseSharingInvitationV1(decodeJson(encoded, "invitation"));
}

export function encodeSharingPublicKeyResponseV1(
  response: SharingPublicKeyResponseV1,
): string {
  return encodeJson(parseSharingPublicKeyResponseV1(response));
}

export function decodeSharingPublicKeyResponseV1(
  encoded: string,
): SharingPublicKeyResponseV1 {
  return parseSharingPublicKeyResponseV1(decodeJson(encoded, "key response"));
}

export function encodeSharingDatasetFilesV1(
  files: SharingDatasetFileV1[],
): string {
  if (files.length === 0) {
    throw new SyncKitError("compatibility", "A join link needs at least one dataset file.");
  }
  return encodeJson(files.map(normalizeDatasetFile));
}

export function decodeSharingDatasetFilesV1(
  encoded: string,
): SharingDatasetFileV1[] {
  const value = decodeJson(encoded, "dataset files");
  if (!Array.isArray(value) || value.length === 0) {
    throw new SyncKitError("compatibility", "The join link dataset file list is malformed.");
  }
  return value.map(normalizeDatasetFile);
}

// --- link builders / parsers ---

/**
 * Builds a join link carrying the signed invitation and the recipient's granted
 * dataset files. Keeps the legacy folder/exchange/marker params so existing
 * deep-link handlers still recognize it as a join link.
 */
export function buildSharingJoinLinkV1(input: {
  landingUrl: string;
  invitation: SharingInvitationV1;
  files: SharingDatasetFileV1[];
}): string {
  const url = new URL(input.landingUrl);
  url.searchParams.set(SHARING_JOIN_MARKER_PARAM, "1");
  url.searchParams.set(SHARING_JOIN_FOLDER_PARAM, input.invitation.appFolderId);
  url.searchParams.set(SHARING_JOIN_EXCHANGE_PARAM, input.invitation.exchangeId);
  url.searchParams.set(
    SHARING_JOIN_INVITATION_PARAM,
    encodeSharingInvitationV1(input.invitation),
  );
  url.searchParams.set(
    SHARING_JOIN_FILES_PARAM,
    encodeSharingDatasetFilesV1(input.files),
  );
  return url.toString();
}

export function parseSharingJoinLinkV1(
  input: string | URLSearchParams,
): SharingJoinLinkV1 | null {
  const params = toSearchParams(input);
  const invitation = params.get(SHARING_JOIN_INVITATION_PARAM);
  const files = params.get(SHARING_JOIN_FILES_PARAM);
  if (!invitation || !files) return null;
  return {
    invitation: decodeSharingInvitationV1(invitation),
    files: decodeSharingDatasetFilesV1(files),
  };
}

/** Builds a response link the recipient sends back to the owner to finish join. */
export function buildSharingResponseLinkV1(input: {
  landingUrl: string;
  response: SharingPublicKeyResponseV1;
}): string {
  const url = new URL(input.landingUrl);
  url.searchParams.set(SHARING_RESPONSE_MARKER_PARAM, "1");
  url.searchParams.set(SHARING_JOIN_EXCHANGE_PARAM, input.response.exchangeId);
  url.searchParams.set(
    SHARING_RESPONSE_PAYLOAD_PARAM,
    encodeSharingPublicKeyResponseV1(input.response),
  );
  return url.toString();
}

export function parseSharingResponseLinkV1(
  input: string | URLSearchParams,
): SharingResponseLinkV1 | null {
  const params = toSearchParams(input);
  if (params.get(SHARING_RESPONSE_MARKER_PARAM) !== "1") return null;
  const response = params.get(SHARING_RESPONSE_PAYLOAD_PARAM);
  if (!response) return null;
  return { response: decodeSharingPublicKeyResponseV1(response) };
}

// --- internals ---

function encodeJson(value: unknown): string {
  return bytesToBase64Url(new TextEncoder().encode(JSON.stringify(value)));
}

function decodeJson(encoded: string, label: string): unknown {
  let text: string;
  try {
    text = new TextDecoder().decode(base64UrlToBytes(encoded));
  } catch (error) {
    throw new SyncKitError("compatibility", `The ${label} payload is not valid base64url.`, {
      cause: error,
    });
  }
  try {
    return JSON.parse(text) as unknown;
  } catch (error) {
    throw new SyncKitError("compatibility", `The ${label} payload is not valid JSON.`, {
      cause: error,
    });
  }
}

function normalizeDatasetFile(value: unknown): SharingDatasetFileV1 {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new SyncKitError("compatibility", "A dataset file entry must be an object.");
  }
  const entry = value as Record<string, unknown>;
  for (const field of ["datasetId", "fileId", "role"]) {
    const fieldValue = entry[field];
    if (typeof fieldValue !== "string" || fieldValue.length === 0) {
      throw new SyncKitError("compatibility", `A dataset file ${field} must be a non-empty string.`);
    }
  }
  const role = entry.role as string;
  if (role !== "admin" && role !== "writer" && role !== "viewer") {
    throw new SyncKitError("compatibility", `A dataset file has an unsupported role: ${role}.`);
  }
  return {
    datasetId: entry.datasetId as string,
    fileId: entry.fileId as string,
    role,
  };
}

function toSearchParams(input: string | URLSearchParams): URLSearchParams {
  if (input instanceof URLSearchParams) return input;
  if (input.startsWith("?")) return new URLSearchParams(input.slice(1));
  if (input.includes("://")) return new URL(input).searchParams;
  return new URLSearchParams(input);
}

File: android/synckit/src/main/java/com/keyneom/synckit/sharing/LinkExchange.kt
package com.keyneom.synckit.sharing

import com.keyneom.synckit.core.SyncKitError
import com.keyneom.synckit.core.SyncKitErrorCode
import com.keyneom.synckit.crypto.Base64Url
import com.keyneom.synckit.crypto.CanonicalJson
import com.keyneom.synckit.crypto.SyncKitJson
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

/**
 * Link-carried key exchange (Kotlin parity with src/sharing/link-exchange.ts).
 * The signed invitation and key response travel in the join/response links
 * (base64url of the existing structs) rather than Drive `exchanges/` files, so a
 * `drive.file` recipient never reads a file the owner authored and the owner
 * never reads a file the recipient authored. See docs/sync-kit-linkbased-join.md.
 */

const val SHARING_JOIN_INVITATION_PARAM = "sk-inv"
const val SHARING_JOIN_FILES_PARAM = "sk-files"
const val SHARING_RESPONSE_MARKER_PARAM = "sk-resp"
const val SHARING_RESPONSE_PAYLOAD_PARAM = "sk-kr"
private const val SHARING_LINK_PERMISSION_PREFIX = "link:"

/** Legacy unbound marker retained for source compatibility; new links reject it. */
@Deprecated("Use createSharingLinkPermissionIdV1(files)")
const val SHARING_LINK_PERMISSION_ID = "link"

/** A dataset file the recipient is granted, so the Picker can offer it by id. */
@Serializable
data class SharingDatasetFileV1(
    val datasetId: String,
    val fileId: String,
    val role: SharingRole,
)

data class SharingJoinLinkV1(
    val invitation: SharingInvitationV1,
    val files: List<SharingDatasetFileV1>,
)

data class SharingResponseLinkV1(
    val response: SharingPublicKeyResponseV1,
)

/** Owner-side result of a link-carried invite: what to embed in the join link. */
data class SharingLinkInvite(
    val invitation: SharingInvitationV1,
    val files: List<SharingDatasetFileV1>,
)

// --- payload encode/decode (signed structs unchanged; only transported) ---

fun encodeSharingInvitationV1(invitation: SharingInvitationV1): String =
    encodeJson(
        SharingCrypto.parseSharingInvitationV1(invitation),
        SharingInvitationV1.serializer(),
    )

fun decodeSharingInvitationV1(encoded: String): SharingInvitationV1 =
    SharingCrypto.parseSharingInvitationV1(
        decodeJson(encoded, "invitation", SharingInvitationV1.serializer()),
    )

fun encodeSharingPublicKeyResponseV1(response: SharingPublicKeyResponseV1): String =
    encodeJson(
        SharingCrypto.parseSharingPublicKeyResponseV1(response),
        SharingPublicKeyResponseV1.serializer(),
    )

fun decodeSharingPublicKeyResponseV1(encoded: String): SharingPublicKeyResponseV1 =
    SharingCrypto.parseSharingPublicKeyResponseV1(
        decodeJson(encoded, "key response", SharingPublicKeyResponseV1.serializer()),
    )

fun encodeSharingDatasetFilesV1(files: List<SharingDatasetFileV1>): String {
    return encodeJson(normalizedDatasetFiles(files), ListSerializer(SharingDatasetFileV1.serializer()))
}

fun decodeSharingDatasetFilesV1(encoded: String): List<SharingDatasetFileV1> {
    val files = decodeJson(encoded, "dataset files", ListSerializer(SharingDatasetFileV1.serializer()))
    if (files.isEmpty()) {
        throw SyncKitError(SyncKitErrorCode.COMPATIBILITY, "The join link dataset file list is malformed.")
    }
    return normalizedDatasetFiles(files)
}

/** Binds the file manifest into the invitation's signed permission field. */
fun createSharingLinkPermissionIdV1(files: List<SharingDatasetFileV1>): String {
    val normalized = normalizedDatasetFiles(files)
    val payload = buildJsonArray {
        normalized.forEach { file ->
            add(
                buildJsonObject {
                    put("datasetId", file.datasetId)
                    put("fileId", file.fileId)
                    put(
                        "role",
                        when (file.role) {
                            SharingRole.ADMIN -> "admin"
                            SharingRole.WRITER -> "writer"
                            SharingRole.VIEWER -> "viewer"
                            SharingRole.OWNER -> error("Owner is not a supported link grant role.")
                        },
                    )
                },
            )
        }
    }
    val digest = MessageDigest.getInstance("SHA-256").digest(CanonicalJson.encodeAad(payload))
    return SHARING_LINK_PERMISSION_PREFIX + Base64Url.encode(digest)
}

/** Verifies exact signed-grant/file-manifest correspondence before registry use. */
fun verifySharingLinkDatasetFilesV1(
    invitation: SharingInvitationV1,
    files: List<SharingDatasetFileV1>,
): List<SharingDatasetFileV1> {
    val normalized = normalizedDatasetFiles(files)
    val grants = invitation.requestedGrants.sortedWith { left, right ->
        CanonicalJson.compareUtf16CodeUnits(left.datasetId, right.datasetId)
    }
    if (
        normalized.size != grants.size ||
        normalized.indices.any { index ->
            normalized[index].datasetId != grants[index].datasetId ||
                normalized[index].role != grants[index].role
        }
    ) {
        throw SyncKitError(
            SyncKitErrorCode.AUTHORIZATION,
            "The sharing link file manifest does not exactly match its signed grants.",
        )
    }
    if (invitation.recipientDrivePermissionId != createSharingLinkPermissionIdV1(normalized)) {
        throw SyncKitError(
            SyncKitErrorCode.AUTHORIZATION,
            "The sharing link file manifest is not authenticated by its invitation.",
        )
    }
    return normalized
}

// --- link builders / parsers ---

fun buildSharingJoinLinkV1(
    landingUrl: String,
    invitation: SharingInvitationV1,
    files: List<SharingDatasetFileV1>,
): String = appendQueryParams(
    landingUrl,
    linkedMapOf(
        SHARING_JOIN_MARKER_PARAM to "1",
        SHARING_JOIN_FOLDER_PARAM to invitation.appFolderId,
        SHARING_JOIN_EXCHANGE_PARAM to invitation.exchangeId,
        SHARING_JOIN_INVITATION_PARAM to encodeSharingInvitationV1(invitation),
        SHARING_JOIN_FILES_PARAM to encodeSharingDatasetFilesV1(files),
    ),
)

fun parseSharingJoinLinkV1(input: String): SharingJoinLinkV1? {
    val params = parseQuery(input)
    val invitation = params[SHARING_JOIN_INVITATION_PARAM]
    val files = params[SHARING_JOIN_FILES_PARAM]
    if (invitation.isNullOrEmpty() || files.isNullOrEmpty()) return null
    return SharingJoinLinkV1(
        invitation = decodeSharingInvitationV1(invitation),
        files = decodeSharingDatasetFilesV1(files),
    )
}

fun buildSharingResponseLinkV1(
    landingUrl: String,
    response: SharingPublicKeyResponseV1,
): String = appendQueryParams(
    landingUrl,
    linkedMapOf(
        SHARING_RESPONSE_MARKER_PARAM to "1",
        SHARING_JOIN_EXCHANGE_PARAM to response.exchangeId,
        SHARING_RESPONSE_PAYLOAD_PARAM to encodeSharingPublicKeyResponseV1(response),
    ),
)

fun parseSharingResponseLinkV1(input: String): SharingResponseLinkV1? {
    val params = parseQuery(input)
    if (params[SHARING_RESPONSE_MARKER_PARAM] != "1") return null
    val response = params[SHARING_RESPONSE_PAYLOAD_PARAM]
    if (response.isNullOrEmpty()) return null
    return SharingResponseLinkV1(decodeSharingPublicKeyResponseV1(response))
}

// --- internals ---

private fun <T> encodeJson(value: T, serializer: kotlinx.serialization.KSerializer<T>): String =
    Base64Url.encode(
        SyncKitJson.instance.encodeToString(serializer, value).toByteArray(StandardCharsets.UTF_8),
    )

private fun <T> decodeJson(
    encoded: String,
    label: String,
    serializer: kotlinx.serialization.KSerializer<T>,
): T {
    val json = try {
        String(Base64Url.decode(encoded), StandardCharsets.UTF_8)
    } catch (error: Exception) {
        throw SyncKitError(SyncKitErrorCode.COMPATIBILITY, "The $label payload is not valid base64url.", error)
    }
    return try {
        SyncKitJson.instance.decodeFromString(serializer, json)
    } catch (error: Exception) {
        throw SyncKitError(SyncKitErrorCode.COMPATIBILITY, "The $label payload is not valid JSON.", error)
    }
}

private fun normalizeDatasetFile(file: SharingDatasetFileV1): SharingDatasetFileV1 {
    if (file.datasetId.isBlank() || file.fileId.isBlank()) {
        throw SyncKitError(SyncKitErrorCode.COMPATIBILITY, "A dataset file id must be a non-empty string.")
    }
    if (file.role == SharingRole.OWNER) {
        throw SyncKitError(SyncKitErrorCode.COMPATIBILITY, "A dataset file has an unsupported role: owner.")
    }
    return file
}

private fun normalizedDatasetFiles(files: List<SharingDatasetFileV1>): List<SharingDatasetFileV1> {
    if (files.isEmpty()) {
        throw SyncKitError(SyncKitErrorCode.COMPATIBILITY, "A join link needs at least one dataset file.")
    }
    val normalized = files.map(::normalizeDatasetFile).sortedWith { left, right ->
        CanonicalJson.compareUtf16CodeUnits(left.datasetId, right.datasetId)
    }
    normalized.zipWithNext().firstOrNull { (left, right) -> left.datasetId == right.datasetId }?.let {
        throw SyncKitError(
            SyncKitErrorCode.COMPATIBILITY,
            "Duplicate dataset file ${it.first.datasetId}.",
        )
    }
    return normalized
}

private fun appendQueryParams(landingUrl: String, params: Map<String, String>): String {
    val uri = URI(landingUrl)
    val query = linkedMapOf<String, String>()
    uri.rawQuery?.split("&")?.forEach { pair ->
        val parts = pair.split("=", limit = 2)
        if (parts.isNotEmpty() && parts[0].isNotBlank()) {
            query[parts[0]] = parts.getOrNull(1)
                ?.let { URLDecoder.decode(it, StandardCharsets.UTF_8.name()) }
                .orEmpty()
        }
    }
    query.putAll(params)
    val queryString = query.entries.joinToString("&") { (key, value) ->
        "$key=${URLEncoder.encode(value, StandardCharsets.UTF_8.name())}"
    }
    val path = uri.rawPath.orEmpty()
    val fragment = uri.rawFragment?.let { "#$it" }.orEmpty()
    return "${uri.scheme}://${uri.authority}$path?$queryString$fragment"
}

private fun parseQuery(input: String): Map<String, String> {
    val raw = when {
        input.startsWith("?") -> input.drop(1)
        "://" in input -> URI(input).rawQuery.orEmpty()
        else -> input
    }
    if (raw.isBlank()) return emptyMap()
    return raw.split("&").mapNotNull { pair ->
        val parts = pair.split("=", limit = 2)
        if (parts[0].isBlank()) null else {
            parts[0] to parts.getOrNull(1)
                ?.let { URLDecoder.decode(it, StandardCharsets.UTF_8.name()) }
                .orEmpty()
        }
    }.toMap()
}
